### PR TITLE
[nat64] add function for parsing IPv4 CIDR

### PR DESCRIPTION
**Pull request by erjiaqing**

### CODE REVIEW

File: src/core/net/ip4_types.cpp [hunk 1]
```diff
@@ -148,6 +147,46 @@ Address::InfoString Address::ToString(void) const
...
+        {
+            if ((digitChar < '0') || (digitChar > '9'))
+            {
+                break;
+            }
+
+            value = static_cast<uint16_t>((value * 10) + static_cast<uint8_t>(digitChar - '0'));
+            VerifyOrExit(value <= kMaxCidrLength);
+            hasFirstDigit = true;
+        }
```
Comment (abtink) [inline src/core/net/ip4_types.cpp:177]: We has similar code in `Address::FromString()` reading an int value?
Maybe we can define a helper method/function for this (in `string.hpp` maybe so even more general and can be used by all modules)?

File: tests/unit/test_ip_address.cpp [hunk 2]
```diff
@@ -194,6 +194,77 @@ void TestIp4AddressFromString(void)
...
+        {"1.2.3.4", {0}, 0, ot::kErrorParse},          // Where is the suffix?
+        // invalid address and invalid suffix
+        {"123.231.0.256/41", {0}, 0, ot::kErrorParse},     // Invalid byte value.
+        {"100123.231.0.256/abc", {0}, 0, ot::kErrorParse}, // Invalid byte value.
+        {"1.22.33", {0}, 0, ot::kErrorParse},              // Too few bytes.
+        {"1.22.33.44.5/36", {0}, 0, ot::kErrorParse},      // Too many bytes.
+        {"a.b.c.d/99", {0}, 0, ot::kErrorParse},           // Wrong digit char.
+        {"123.23.45 .12", {0}, 0, ot::kErrorParse},        // Extra space.
+        {".", {0}, 0, ot::kErrorParse},                    // Invalid.
+    };
```
Comment (abtink) [inline tests/unit/test_ip_address.cpp:260]: Nice! 👍  Thanks for covering all edge-cases.

**Review by abtink — APPROVED** [review]
Looks good. 
Just one suggestion below (can be done maybe in future PR?)

File: src/core/net/ip4_types.cpp (hunk 1 above)
Comment (erjiaqing) [reply src/core/net/ip4_types.cpp:177]: Would expect another PR for this.

File: src/core/net/ip4_types.cpp [hunk 3]
```diff
@@ -37,10 +37,9 @@
 namespace ot {
 namespace Ip4 {
 
-Error Address::FromString(const char *aString)
+Error Address::FromString(const char *aString, char aTerminalChar)
```
Comment (jwhui) [inline src/core/net/ip4_types.cpp:40]:
```suggestion
Error Address::FromString(const char *aString, char aTerminatorChar)
```

File: src/core/net/ip4_types.hpp [hunk 4]
```diff
@@ -148,7 +148,7 @@ class Address : public otIp4Address, public Equatable<Address>, public Clearable
     void SynthesizeFromCidrAndHost(const Cidr &aCidr, uint32_t aHost);
 
     /**
-     * This method parses an IPv4 address string.
+     * This method parses an IPv4 address string terminated by `aTerminalChar`.
```
Comment (jwhui) [inline src/core/net/ip4_types.hpp:151]:
```suggestion
     * This method parses an IPv4 address string terminated by `aTerminatorChar`.
```

File: src/core/net/ip4_types.hpp [hunk 5]
```diff
@@ -159,7 +159,7 @@ class Address : public otIp4Address, public Equatable<Address>, public Clearable
      * @retval kErrorParse        Failed to parse the IPv4 address string.
      *
      */
-    Error FromString(const char *aString);
+    Error FromString(const char *aString, char aTerminalChar = kNullChar);
```
Comment (jwhui) [inline src/core/net/ip4_types.hpp:162]:
```suggestion
    Error FromString(const char *aString, char aTerminatorChar = kNullChar);
```

File: include/openthread/nat64.h [hunk 6]
```diff
@@ -435,6 +435,18 @@ void otIp4AddressToString(const otIp4Address *aAddress, char *aBuffer, uint16_t
 
 #define OT_IP4_CIDR_STRING_SIZE 20 ///< Length of 000.000.000.000/00 plus a suffix NUL
 
+/**
+ * This function converts a human-readable IPv4 CIDR string into a binary representation.
+ *
+ * @param[in]   aString   A pointer to a NULL-terminated string.
+ * @param[out]  aCidr  A pointer to an IPv4 CIDR.
```
Comment (jwhui) [inline include/openthread/nat64.h:442]:
```suggestion
 * @param[in]   aString   A pointer to a NULL-terminated string.
 * @param[out]  aCidr     A pointer to an IPv4 CIDR.
```

**Review by jwhui — APPROVED** [review]
